### PR TITLE
[AIR] Use a configurable ray temp directory for the `TempFileLock` util

### LINKED ISSUE #32776: [Core] Permission denied error when using tune.Tuner() on a shared machine
### What happened + What you expected to happen

Similar to [this issue](https://github.com/ray-project/ray/issues/3899), I run into a permission error when trying to access `/tmp/ray` or `/tmp/_ray_lockfiles` if another user has created the directories.

However, because I use `tune.Tuner()` and do not call `ray.init()` explicitly, the solution proposed in the linked issue does not apply.

Changing permissions of the said directories to 777 does not resolve the issue.

A band-aid solution is to ask a user with sudo permission to remove the directories every time the error shows up.

### Versions / Dependencies

Ray 2.2.0
Python 3.10.8
PyTorch 1.13.1
Ubuntu 22.04

### Reproduction script

N/A

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by justinvyu**
This PR switches the temp directory used by the `TempFileLock` util to be one that's configurable via environment variable.

## Why are these changes needed?

AIR Checkpoints use a `TempFileLock` utility internally, and it saved these file locks to a `/tmp/_ray_lockfiles` temp directory that could not be configured. This leads to a permission denied error when trying to run Tune experiment on a machine shared with other users.

After this PR, users can do:

```python
ray.init(runtime_env={"env_vars": {"RAY_TMPDIR": "/tmp/abcdef"}}, _temp_dir="/tmp/ray_tmp_dir")
results = Tuner(...).fit()
# Temp file locks are created at `/tmp/abcdef/ray`
```

**Question**: Why are `ray.init(_temp_dir)` and `ray._private.utils.get_ray_temp_dir()` different?

## Related issue number

Closes https://github.com/ray-project/ray/issues/32776

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — APPROVED** [review]

File: python/ray/air/_internal/filelock.py [hunk 1]
```diff
@@ -14,7 +14,8 @@ class TempFileLock:
 
     def __init__(self, path: str, **kwargs):
         self.path = path
-        self._lock_dir = Path(tempfile.gettempdir()).resolve() / RAY_LOCKFILE_DIR
+        temp_dir = Path(ray._private.utils.get_ray_temp_dir()).resolve()
```
Comment (matthewdeng) [inline python/ray/air/_internal/filelock.py:17]: > ray.init(runtime_env={"env_vars": {"TMPDIR": "/tmp/abcdef"}}, _temp_dir="/tmp/ray_tmp_dir")

What would the temp directory be in this case?

Comment (justinvyu) [reply python/ray/air/_internal/filelock.py:17]: Actually, it needs to be:

```
ray.init(runtime_env={"env_vars": {"RAY_TMPDIR": "/tmp/abcdef"}}, _temp_dir="/tmp/ray_tmp_dir")
```

to work on all platforms. `TMPDIR` only applies to linux for some reason.

In this case, `ray._private.utils.get_ray_temp_dir() == "/tmp/abcdef"`.

Comment (justinvyu) [reply python/ray/air/_internal/filelock.py:17]: See here: https://github.com/ray-project/ray/blob/9acf5832246762cc034d37e3926cb31da7c8fde4/python/ray/_private/utils.py#L73

Comment (matthewdeng) [reply python/ray/air/_internal/filelock.py:17]: Oh, what is the `_temp_dir="/tmp/ray_tmp_dir"` arg for?

Comment (matthewdeng) [reply python/ray/air/_internal/filelock.py:17]: @jjyao do you know what the difference is between `RAY_TMPDIR` and `_temp_dir`?

Comment (justinvyu) [reply python/ray/air/_internal/filelock.py:17]: The difference between `ray._private.utils.get_ray_temp_dir` and `ray.init(_temp_dir)`.

Comment (jjyao) [reply python/ray/air/_internal/filelock.py:17]: oh, I checked the code and I think it's just our code issue: we have these two paths to set and get temp dir and they are not always consistent. I'll create an issue to track this.

Comment (jjyao) [reply python/ray/air/_internal/filelock.py:17]: https://github.com/ray-project/ray/issues/32962

**Pull request merged**